### PR TITLE
Added LibtorchModel class, test and documentation

### LINKED ISSUE #30461: Add capability to evaluate arbitrary libtorch models using NEML2 in MOOSE
### Motivation

Provide a way to connect MOOSE simulations with libtorch models that have an arbitrary number of inputs mapped to an arbitrary number of outputs. Using a NEML2 material will allow this to run as a tensorized operation over the entire simulation domain and use available accelerators when possible. 

### Design

Create a NEML2 material to evaluate a libtorch neural network model with an arbitrary number of inputs and mapped to an arbitrary number of outputs.

### Impact

Utilizes existing MOOSE-NEML2 coupling but adds new capability to evaluate libtorch models using it in MOOSE.

### PR DESCRIPTION
**Pull request by chaibhave**
Closes #30461

### CODE REVIEW

File: modules/solid_mechanics/neml2/src/LibtorchModel.cxx [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    {
+      auto val = (_inputs[i]->value() - _x_mean[i]) / _x_std[i];
+      values.push_back(val);
+    }
+
+    auto x = Tensor(torch::transpose(torch::vstack(at::ArrayRef<at::Tensor>(
+                                         values.data(), static_cast<int64_t>(values.size()))),
+                                     0,
+                                     1),
+                    _inputs[0]->batch_dim());
```
Comment (dschwen) [inline modules/solid_mechanics/neml2/src/LibtorchModel.cxx:96]: We should probably assert that all inputs have the same batch dim

File: modules/solid_mechanics/neml2/src/LibtorchModel.cxx [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    x.requires_grad_(true);
+    x.retain_grad();
```
Comment (dschwen) [inline modules/solid_mechanics/neml2/src/LibtorchModel.cxx:99]: why?

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LibtorchModel.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# LibtorchModel
+
+!alert note
+This is a custom [NEML2](solid_mechanics/NEML2.md) object for solid_mechanics. To enable it, configure MOOSE with libtorch and initialize the [NEML2 submodule](https://github.com/applied-material-modeling/neml2).
+
+The documentation is for a NEML2 [!citep](neml2osti, neml2_anl_report) material for evaluating libtorch neural networks. This allows evaluating neural networks with an arbitrary number of inputs and mapping them to an arbitrary number of outputs. This material does not currently support backwards propagation to train the neural network, and is only meant for rapid evaluation of pre-trained neural networks.
```
[the remark this answers is not in the document]
Comment (chaibhave) [reply modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LibtorchModel.md:6]: Done.

File: modules/solid_mechanics/neml2/src/LibtorchModel.cxx (hunk 2 above)
Comment (chaibhave) [reply modules/solid_mechanics/neml2/src/LibtorchModel.cxx:99]: It was needed at some earlier point when I was trying manual derivative calculations, removed it now.

File: modules/solid_mechanics/neml2/src/LibtorchModel.cxx (hunk 1 above)
Comment (chaibhave) [reply modules/solid_mechanics/neml2/src/LibtorchModel.cxx:84]: Done on line 72-76 now

File: modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LibtorchModel.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+## Example Input File
+
+Running NEML2 models using MOOSE requires a separate NEML2 input file that builds the material. An example input for a NEML2 LibtorchModel is shown below. The inputs to the libtorch model are provided using `inputs = 'forces/T'`, and the model outputs are mapped to `outputs = 'state/k_T'`. Since neural networks are typically trained on scaled inputs, the user is also required to specify the `x_mean`,`x_std` and `y_mean`,`y_std` for the input and outputs to be scaled linearly.
```
[the remark this answers is not in the document]
Comment (chaibhave) [reply modules/solid_mechanics/doc/content/modules/solid_mechanics/common/LibtorchModel.md]: Done.

File: modules/solid_mechanics/neml2/include/LibtorchModel.h [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+#include "neml2/models/Model.h"
+#include <torch/script.h>
+#include "DataFileUtils.h"
+namespace neml2
+{
+
+/**
+ * Compute a doped UO2 creep flow rate using an ANN surrogate model.
+ * This does not include irradiation creep.
+ */
```
[the remark this answers is not in the document]
Comment (chaibhave) [reply modules/solid_mechanics/neml2/include/LibtorchModel.h]: Done.

**Review by dschwen — APPROVED** [review]
Looks good to me
